lib/rpc/src/middleware.rs

```diff
@@ -0,0 +1,99 @@
+use std::time::{Duration, Instant};
+
+use crate::metrics::API_METRICS;
+use jsonrpsee::core::middleware::{Batch, Notification};
+use jsonrpsee::server::middleware::rpc::{RpcService, RpcServiceT};
+use jsonrpsee::types::Request;
+
+#[derive(Debug)]
+pub enum CallKind {
+    Call,
+    Notification,
+    Batch,
+}
+
+#[derive(Clone)]
+pub struct Monitoring {
+    inner: RpcService,
+}
+
+impl Monitoring {
+    pub fn new(inner: RpcService) -> Self {
+        Self { inner }
+    }
+}
+
+impl RpcServiceT for Monitoring {
+    type MethodResponse = <RpcService as RpcServiceT>::MethodResponse;
+    type NotificationResponse = <RpcService as RpcServiceT>::NotificationResponse;
+    type BatchResponse = <RpcService as RpcServiceT>::BatchResponse;
+
+    fn call<'a>(
+        &self,
+        request: Request<'a>,
+    ) -> impl Future<Output = Self::MethodResponse> + Send + 'a {
+        let method = request.method_name().to_owned();
+        let fut = self.inner.call(request);
+
+        async move {
+            let started = Instant::now();
+            let out = fut.await;
+            let output_size = out.as_json().get().len();
+
+            log_and_report(CallKind::Call, &method, started.elapsed(), output_size);
```

Comment:
can we also track the request size? Not overly important, may be handy if there is some malicious activity involving very large requests?
So I'd say only add it if it's trivial to add